JavaSource/org/unitime/timetable/server/courses/CourseOfferingPropertiesBackend.java

```diff
@@ -0,0 +1,411 @@
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.CourseTypeInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.DepartmentInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.InstructorInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.OfferingConsentTypeInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.OverrideTypeInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.ResponsibilityInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.SubjectAreaInterface;
+import org.unitime.timetable.gwt.shared.CourseOfferingInterface.WaitListInterface;
+import org.unitime.timetable.model.CourseCreditFormat;
+import org.unitime.timetable.model.CourseCreditType;
+import org.unitime.timetable.model.CourseCreditUnitType;
+import org.unitime.timetable.model.CourseOffering;
+import org.unitime.timetable.model.CourseType;
+import org.unitime.timetable.model.Department;
+import org.unitime.timetable.model.DepartmentalInstructor;
+import org.unitime.timetable.model.InstructionalOffering;
+import org.unitime.timetable.model.OfferingConsentType;
+import org.unitime.timetable.model.OverrideType;
+import org.unitime.timetable.model.Session;
+import org.unitime.timetable.model.SubjectArea;
+import org.unitime.timetable.model.TeachingResponsibility;
+import org.unitime.timetable.model.comparators.CourseOfferingComparator;
+import org.unitime.timetable.model.dao.CourseTypeDAO;
+import org.unitime.timetable.model.dao.DepartmentDAO;
+import org.unitime.timetable.model.dao.DepartmentalInstructorDAO;
+import org.unitime.timetable.model.dao.OverrideTypeDAO;
+import org.unitime.timetable.model.dao.SessionDAO;
+import org.unitime.timetable.model.dao.SubjectAreaDAO;
+import org.unitime.timetable.security.SessionContext;
+import org.unitime.timetable.security.UserAuthority;
+import org.unitime.timetable.security.rights.Right;
+import org.unitime.timetable.util.Constants;
+
+/**
+ * @author Alec Macleod
+ */
+@GwtRpcImplements(CourseOfferingPropertiesRequest.class)
+public class CourseOfferingPropertiesBackend implements GwtRpcImplementation<CourseOfferingPropertiesRequest, CourseOfferingPropertiesInterface> {
+
+	protected final static CourseMessages MSG = Localization.create(CourseMessages.class);
+	Logger logger = java.util.logging.Logger.getLogger("UpdateCourseOfferingBackend");
+
+	@Override
+	public CourseOfferingPropertiesInterface execute(CourseOfferingPropertiesRequest request, SessionContext context) {
+		
+		if (request.hasSessionId())
+			context = new EventContext(context, request.getSessionId());
+
+		if (!context.hasPermission(Right.AddCourseOffering) && !context.hasPermission(Right.EditCourseOffering) && !context.hasPermission(Right.EditCourseOfferingCoordinators) && !context.hasPermission(Right.EditCourseOfferingNote)) {
```

Comment:
This would not work. The context.hasPermission(Right.AddCourseOffering) is somewhat fine -- it returns true if there is at least one subject area for which the user has the AddCourseOffering permission. The other three permission checks will always be false -- there is no course offering provided they can be evaluated on!

How about something like:
```
if (request.getCourseOfferingId() != null) { // edit -- must have at least one edit permission
    if (!context.hasPermission(request.getCourseOfferingId(), Right.EditCourseOffering) &&
        !context.hasPermission(request.getCourseOfferingId(), Right.EditCourseOfferingCoordinators) &&
        !context.hasPermission(request.getCourseOfferingId(), Right.EditCourseOfferingNote)) {
        context.checkPermission(request.getCourseOfferingId(), Right.EditCourseOffering);
    }
} else { // check Add permission otherwise (the user has at least one subject area for which he/she can add a course)
    context.checkPermission(Right.AddCourseOffering);
}
```